### PR TITLE
Migrated Blob Batch to the new generator.

### PR DESCRIPTION
**Pull request by seanmcc-msft**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs [hunk 1]
```diff
@@ -153,15 +191,36 @@ private void SetBatchOperationType(BlobBatchOperationType operationType)
...
+            HttpMessage message = BlobRestClient.CreateDeleteRequest(
+                containerName: blobContainerName,
+                blob: blobName,
+                timeout: null,
+                leaseId: conditions?.LeaseId,
+                deleteSnapshots: snapshotsOption == DeleteSnapshotsOption.None ? null : (DeleteSnapshotsOptionType?)snapshotsOption,
+                ifModifiedSince: conditions?.IfModifiedSince,
+                ifUnmodifiedSince: conditions?.IfUnmodifiedSince,
+                ifMatch: conditions?.IfMatch.ToString(),
+                ifNoneMatch: conditions?.IfNoneMatch.ToString(),
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs:205]: can IfMatch/IfNoneMatch be null ?

File: sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs [hunk 2]
```diff
@@ -153,15 +191,36 @@ private void SetBatchOperationType(BlobBatchOperationType operationType)
...
+                message,
+                async response =>
+                {
+                    switch (response.Status)
+                    {
+                        case 202:
+                            BlobDeleteHeaders blobDeleteHeaders = new BlobDeleteHeaders(response);
+                            return ResponseWithHeaders.FromValue(blobDeleteHeaders, response);
+                        default:
+                            throw await _clientDiagnostics.CreateRequestFailedExceptionAsync(response).ConfigureAwait(false);
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs:221]: when this code is executed? would it make public API throw or would errored message be in enumeration in the API result?
Worth checking if we have existing test that hits such logical branch. If we have one then we're good, if we don't we should get one.
(similar question elsewhere, it seems to form a pattern)

File: sdk/storage/Azure.Storage.Blobs.Batch/src/ContainerRestClient.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+using System.IO;
+using Azure.Core;
+
+namespace Azure.Storage.Blobs.Batch
+{
+    internal partial class ContainerRestClient
+    {
+        // We are overriding this method because the new generator attempts to add 2 content types to the Content-Type header,
+        // causing auth to fail.  https://github.com/Azure/autorest/issues/3914
+        internal HttpMessage CreateSubmitBatchRequest(string containerName, long contentLength, string multipartContentType, Stream body, int? timeout)
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.Batch/src/ContainerRestClient.cs:14]: Could you please create issue in our repo that tracks that as well? We should be able to remove this customization when generator work is done.

File: sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/BlobDeleteType.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        /// <summary>
+        /// None.
+        /// </summary>
+        None,
+
+        /// <summary>
+        /// Permanent.
+        /// </summary>
+        Permanent
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/BlobDeleteType.cs:16]: was there more verbose doc for this before?

File: sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/DeleteSnapshotsOptionType.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        /// <summary>
+        /// Include.
+        /// </summary>
+        Include,
+
+        /// <summary>
+        /// Only.
+        /// </summary>
+        Only
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/DeleteSnapshotsOptionType.cs:16]: was there more verbose doc for this before?

File: sdk/storage/Azure.Storage.Blobs.Batch/src/ServiceRestClient.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+using System.IO;
+using Azure.Core;
+
+namespace Azure.Storage.Blobs.Batch
+{
+    internal partial class ServiceRestClient
+    {
+        // We are overriding this method because the new generator attempts to add 2 content types to the Content-Type header,
+        // causing auth to fail.  https://github.com/Azure/autorest/issues/3914
+        internal HttpMessage CreateSubmitBatchRequest(long contentLength, string multipartContentType, Stream body, int? timeout)
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.Batch/src/ServiceRestClient.cs:14]: same here. plz reference issue created in our repo in the comments.

**Review by kasobol-msft — COMMENTED** [review]
Overall looking good.

File: sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs:205]: Fixed.

File: sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.Batch/src/BlobBatch.cs:221]: This code is executed when a sub-request fails, an actual exception is not thrown, the error message is added to the result.

There are multiple unit tests (basically any unit tests with the word "fail" in the name) that cover this path, hence how I found I needed to add this.

File: sdk/storage/Azure.Storage.Blobs.Batch/src/ContainerRestClient.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.Batch/src/ContainerRestClient.cs:14]: fixed.

File: sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/BlobDeleteType.cs (hunk 4 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/BlobDeleteType.cs:16]: This is an internal class that customers cannot access.

File: sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/DeleteSnapshotsOptionType.cs (hunk 5 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.Batch/src/Models/Internal/DeleteSnapshotsOptionType.cs:16]: This is an internal class that customers cannot access.

File: sdk/storage/Azure.Storage.Blobs.Batch/src/ServiceRestClient.cs (hunk 6 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.Batch/src/ServiceRestClient.cs:14]: fixed.

**Review by kasobol-msft — APPROVED** [review]

**Pull request merged**